### PR TITLE
Shotgun spread refactor

### PR DESCRIPTION
**Pull request by RiceMar1244**
## About the PR

Makes the code for firing a spread of projectiles use a new component instead of the CartridgeAmmoComponent.

## Why / Balance

I noticed multiple people had tried to introduce energy spread weapons to the game in the past but they were rejected on the basis of spread code.
Previously only shotgun cartridges could define projectile spreads. With this refactor, now things like energy weapons will be able to fire spreads.

## Technical details

I added a new component, ProjectileSpreadComponent. The new component now holds the projectile count and spread data that was previously in the CartridgeAmmoComponent, as well as an additional entity prototype field. In the GunSystem, the code for spawning the original bullet uses the switch case's preferred method (spawning the bullet from the cartridge prototype for CartridgeAmmoComponent and just launching the projectile it already has for AmmoComponent) but uses the ProjectileSpreadComponent to launch the rest of the spread.

## Media

- [ ] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

Nothing breaks as a result of the changes.

**Changelog**

No CL!

### CODE REVIEW

File: Content.Server/Weapons/Ranged/Systems/GunSystem.cs [hunk 1]
```diff
@@ -113,8 +113,11 @@ private void OnBallisticPrice(EntityUid uid, BallisticAmmoProviderComponent comp
 
         foreach (var (ent, shootable) in ammo)
         {
+            if (ent == null)
+                continue;
+
```
[the remark this answers is not in the document]
Comment (RiceMar1244) [reply Content.Server/Weapons/Ranged/Systems/GunSystem.cs:118]: Oh no, thanks for the catch. If I remember correctly, I think the code was complaining about a potentially null value, so I thought id check it at the very start. Ill look into this.

File: Content.Server/Weapons/Ranged/Systems/GunSystem.cs [hunk 2]
```diff
@@ -174,10 +177,36 @@ private void OnBallisticPrice(EntityUid uid, BallisticAmmoProviderComponent comp
...
+                            ShootOrThrow(newuid, angles[i].ToVec(), gunVelocity, gun, gunUid, user);
+                            shotProjectiles.Add(ent!.Value);
+                        }
+                    }
+                    else
+                    {
+                        ShootOrThrow(ent.Value, mapDirection, gunVelocity, gun, gunUid, user);
+                        shotProjectiles.Add(ent!.Value);
+                    }
+
```
Comment (metalgearsloth) [inline Content.Server/Weapons/Ranged/Systems/GunSystem.cs:206]: This is slightly too copy-paste with the above for my taste and needs combining please.

File: Content.Shared/Projectiles/ProjectileSpreadComponent.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    /// <summary>
+    /// How much the ammo spreads when shot, in degrees. Does nothing if count is 0.
+    /// </summary>
+    [ViewVariables(VVAccess.ReadWrite), DataField("spread")]
+    public Angle Spread = Angle.FromDegrees(5);
+
+    /// <summary>
+    /// How many prototypes are spawned when shot.
+    /// </summary>
+    [ViewVariables(VVAccess.ReadWrite), DataField("count")]
```
Comment (metalgearsloth) [inline Content.Shared/Projectiles/ProjectileSpreadComponent.cs:29]: Needs updating to modern code standards i.e.
- Needs to actually be networked.
- EntityPrototype serializer was long deprecated and replaced by entprotoid.
- Datafield strings are unnecessary.
- VVRW is implicit on datafields now.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Comment by RiceMar1244** [comment]
Alrighty, Ill see what I can do. Thanks for the feedback!

**Comment by metalgearsloth** [comment]
Re-open if you come back to it.

**Comment by RiceMar1244** [comment]
Thanks for closing it up for me Sloth. I was hoping Id feel like coming back to it but i dont think I will right now. I think I just really want to work on something of my own right now. Cheers!

**Pull request merged**